File: TypeScript/PaymentsService/src/Infrastructure/Payments/PayPal/PayPalPaymentClient.ts
import * as uuid from 'uuid';
import { IPaymentClient } from '../../../Domain/Payments/IPaymentClient';
import { ProcessPaymentBaseRequest } from '../../../Domain/Payments/ProcessPaymentBaseRequest';
import { ProcessPayPalPaymentRequest } from './ProcessPayPalPaymentRequest';
import { ProcessPayPalPaymentResponse } from './ProcessPayPalPaymentResponse';

export class PayPalPaymentClient implements IPaymentClient {
  processPayment(request: ProcessPaymentBaseRequest): ProcessPayPalPaymentResponse {
    const req = request as ProcessPayPalPaymentRequest;

    if (!this.validateRequest(req)) {
      throw new Error('Invalid request');
    }

    const resp: ProcessPayPalPaymentResponse = new ProcessPayPalPaymentResponse();
    resp.email =`${req.token}@paypal.com`;
    resp.transactionId = uuid.v4();

    if (req.token === 'fail') {
      resp.status = 'Failed';
      resp.error = 'Generic error';
    } else {
      resp.status = 'Success';
    }

    return resp;
  }

  private validateRequest(request: ProcessPayPalPaymentRequest): boolean {
    if (!request) {
      return false;
    }

    if (!request.token) {
      return false;
    }

    return true;
  }
}

File: Python/PaymentsService/Infrastructure/Payments/PayPal/PayPalPaymentClient.py
import uuid

from Domain.Payments.IPaymentClient import IPaymentClient
from Domain.Payments.ProcessPaymentBaseRequest import ProcessPaymentBaseRequest
from Infrastructure.Payments.PayPal.ProcessPayPalPaymentRequest import ProcessPayPalPaymentRequest
from Infrastructure.Payments.PayPal.ProcessPayPalPaymentResponse import ProcessPayPalPaymentResponse

class PayPalPaymentClient(IPaymentClient):
    def process_payment(self, request: ProcessPaymentBaseRequest) -> ProcessPayPalPaymentResponse:
        if not isinstance(request, ProcessPayPalPaymentRequest):
            raise ValueError("Invalid request type")

        resp = ProcessPayPalPaymentResponse()
        resp.email = f"{request.token}@paypal.com"
        resp.transactionId = str(uuid.uuid4())

        if request.token == "fail":
            resp.status = "Failed"
            resp.error = "Generic error"
        else:
            resp.status = "Success"

        return resp